### PR TITLE
feat(moe): add remaining CUTLASS unified MoE runners

### PR DESCRIPTION
**Pull request by feih-nv**
## 📌 Description

Adds the remaining CUTLASS fused-MoE paths to the unified MoE API as **quant-specific** `Cutlass*Config` / `Cutlass*Runner` pairs, on top of the existing BF16 and W4A16 adapters. Each backend has its own weight-prepare helper and activation contract.

`_CutlassRunnerBase` now carries `_x_dtype` and the flat-API flags (`_use_deepseek_fp8_block_scale`, `_use_mxfp8_act_scaling`, `_use_packed_weights`, `_use_wfp4afp8_humming`) so each runner sets those instead of copying the launch path.

Wired backends:

| Config | `QuantVariant` | Arch | Activations |
|---|---|---|---|
| `CutlassNvfp4Config` | `NVFP4` | SM10x / SM110 / SM12x | BF16 (kernel quantizes) |
| `CutlassFp8PerTensorConfig` | `FP8PerTensor` | SM89+ | E4M3 + scalar dequant scale |
| `CutlassFp8BlockConfig` | `DeepSeekFp8` | SM90 | BF16 (kernel quantizes) |
| `CutlassMxfp8Mxfp4Config` | `MXFP4` | SM10x / SM110 / SM12x | MXFP8 + `input_sf` |
| `CutlassMxfp8Config` | `MxFp8` | SM10x (100 / 103 / 107) | MXFP8 + `input_sf` |
| `CutlassW4A8Config` | **new** `W4A8` | SM90 | BF16, packed INT4 |
| `CutlassHummingConfig` | **new** `Humming` | SM90 | BF16, Humming MXFP4 |

The two MXFP8 rows mirror different flat-API skips on purpose: MXFP8xMXFP4 follows `capability[0] not in [10, 11, 12]`, MXFP8xMXFP8 the narrower `not in [10]`.

v1 of each runner is PackedPrecomputed + SwiGLU + `do_finalize=True`, no EP. Legacy `cutlass_fused_moe` tests are unchanged.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

`tests/moe/test_unified_moe_cutlass.py` (127 tests): architecture / registration, prepare-contract rejects, `check_support` rejects, numerical vs independent dequant ref, autotune + CUDA graph. Autotune tests require a concrete `(gemm1, gemm2)` tactic with both IDs `>= 0`.

Local runs:
- **SM100**: full file passes; 5 skipped (SM90-only: W4A16, DeepSeek block FP8, W4A8, Humming)
- **H100 / SM90a**: the 25 tests selected by `-k "w4a16 or fp8_block or w4a8 or humming"` pass (W4A16 numerics + autotune/CUDA graph, DeepSeek block FP8, W4A8, Humming)

## Reviewer Notes

**⚠️ Breaking change.** The deprecated, unregistered `CutlassConfig` placeholder is removed — it was never a runnable `MoELayer` backend (`supported()` always false). Callers must pick a quant-specific type (`CutlassBf16Config`, `CutlassNvfp4Config`, …).

**MXFP8xMXFP8 geometry.** `CutlassMxfp8Config` now requires `hidden_size` and `intermediate_size` divisible by 128 at prepare time (same as MXFP8xMXFP4). Sizes such as `I=64` or `I=192` previously passed prepare and then failed at launch, or read an undersized fc1 scale buffer. That layout never worked; prepare now rejects it instead of waiting for the kernel ICHECK.

**Not in `_DEFAULT_BACKEND`.** These new configs are opt-in: pass an explicit `Cutlass*Config`. `_DEFAULT_BACKEND` is auto-selected, so every candidate must accept the *same* `MoEActivationPack` the caller already prepared (today: the TRTLLM encoding). Auto-picking a CUTLASS runner would hand it a pack it cannot read.

Activation-pack mismatches vs TRTLLM for the same variant:

| Config | CUTLASS activations | TRTLLM activations |
|---|---|---|
| NVFP4 / DeepSeek / W4A8 / Humming | BF16 (kernel quantizes) | already quantized |
| FP8 per-tensor | E4M3 + explicit dequant scale | scale folded into the weight view |
| MXFP8 / MXFP8xMXFP4 | swizzled `input_sf` | linear scales |

`CutlassBf16Config` and `CutlassW4A16Config` stay on the default list because their packs already match.

For the same reason the new configs are **not** added to the unified MoE fuzzer (`#4475`): their packs do not match the existing `_DTYPE` handlers. Say if you want dedicated handlers for `W4A8` / `Humming` in a follow-up.

Please treat these as **separate tensor contracts**, not a universal CUTLASS fallback. Do not mix TRTLLM shuffled / BlockMajorK weights into these prepares.

### CODE REVIEW

**Comment by feih-nv** [comment]
@flashinfer-bot run

File: tests/moe/test_unified_moe_cutlass.py [hunk 1]
```diff
@@ -336,6 +469,188 @@ def test_cutlass_runner_rejects_out_of_scope_configs(config, match):
...
+def test_cutlass_fp8_block_rejects_cuda_below_12_8(monkeypatch):
+    monkeypatch.setattr(
+        "flashinfer.jit.cpp_ext.is_cuda_version_at_least",
+        lambda _version: False,
+    )
+    runner = CutlassFp8BlockRunner.__new__(CutlassFp8BlockRunner)
+    runner.config = _config(quant=QuantConfig(variant=QuantVariant.DeepSeekFp8))
+    runner._device_arch = 90
+    with pytest.raises(NotImplementedError, match="CUDA 12.6 or lower"):
+        runner.check_support()
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply tests/moe/test_unified_moe_cutlass.py:632]: Good catch. The monkeypatch works (the import in check_support is function-local), but the assertion pointed at a wrong message: the gate is is_cuda_version_at_least("12.8"), so 12.7 is rejected too. Fixed in 808b8ac — now "requires CUDA 12.8 or newer", with the flat path aligned in 1553d9a.

File: tests/moe/test_unified_moe_cutlass.py [hunk 2]
```diff
@@ -336,6 +480,240 @@ def test_cutlass_runner_rejects_out_of_scope_configs(config, match):
...
+    runner = CutlassFp8PerTensorRunner.__new__(CutlassFp8PerTensorRunner)
+    hidden = torch.empty(1, 128, dtype=torch.float8_e4m3fn)
+    act = MoEActivationPack(
+        hidden,
+        torch.ones(1, dtype=torch.float32),
+        torch.zeros(1, 2, dtype=torch.int32),
+        torch.ones(1, 2, dtype=torch.float32),
+    )
+    with pytest.raises(ValueError, match="0-dim float32"):
+        runner._validate_activation_scale(act)
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply tests/moe/test_unified_moe_cutlass.py:659]: This one doesn't apply. _use_mxfp8_act_scaling and _x_dtype are ClassVars with class-level defaults on _CutlassRunnerBase (runners.py:395 and :399), and CutlassFp8PerTensorRunner overrides _x_dtype = torch.float8_e4m3fn at :1079. Class attributes are resolved without __init__, so __new__ reaches them and _validate_activation_scale raises the expected ValueError — no AttributeError. The __new__ idiom is used deliberately throughout this file for validation-only tests (see the pre-existing test_cutlass_mxfp8_rejects_linear_activation_scales) to avoid constructing a real runner and touching CUDA. Both tests pass as written.

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

File: flashinfer/fused_moe/api.py [hunk 3]
```diff
@@ -556,31 +580,9 @@ def __repr__(self) -> str:
...
-
-    @classmethod
-    def supported(cls, arch: int) -> bool:
-        # Compatibility-only placeholder: it has no registered runner and must
-        # never be surfaced as a dispatch candidate by BackendOptions.valid_for().
-        return False
-
-    def __repr__(self) -> str:
-        return "CutlassConfig()"
+# Breaking change: the deprecated, unregistered CutlassConfig placeholder was
```
Comment (Aneureka) [inline flashinfer/fused_moe/api.py:583]: Yeah I marked `CutlassConfig` as deprecated earlier, so we can remove it if it is not used. We also don’t need the “Breaking change” comment here.

**Review by Aneureka — APPROVED** [review]
LGTM.

File: flashinfer/fused_moe/api.py (hunk 3 above)
Comment (feih-nv) [reply flashinfer/fused_moe/api.py:583]: Got it. Will update it in https://github.com/flashinfer-ai/flashinfer/pull/4613 as this one is ready to merge.
